### ISSUE TITLE
Add possibility to use include_fields parameter when search with scoped key

### ISSUE THREAD

**Issue by isnigurova-sc** [typesense/typesense#2016]
## Description

There is no possibility to filter the result fields during the search with a scoped key in case the scoped key contains include_fields.
User sees in the documentation that he can use include_fields to list fields from the document to include in the search result. But it doesn't work as expected. The value is ignored, but it should be an intersection with the value from the scoped key.

The same issue is for exclude_fields. It should be a union with the value from the scoped key.

## Steps to reproduce

Prerequisites
1. Collection: "project"
2. A document from the collection has fields "a", "b", "c", "d" and "e".
3. Fields "d"and "e" shouldn't be available for a basic user. A scoped key restrict the output fields using include_fields: a, b, c.

Use case
1. User wants to get field "a" for all projects.
2. User use the scoped key to call typesense.
3. User sends a GET request to /collection/project/document/search with the next params:
    - q: *
    - include_fields: a

## Expected Behavior

User received only field "a" for all projects. 

if user requests fields "a" and "d" then he should receive only field "a", because field "d" is not allowed by scoped key.

## Actual Behavior

User received field "a", "b", "c" for all projects.

## Metadata

**Typesense Version**: 26.0, 27.1

**OS**: doesn't matter

**Comment by jasonbosco** [comment]
The behavior you've described is by design. 

Any parameters you've mentioned in the scoped API Key take precedence over the same parameters mentioned as query parameters. This is because we don't want end-users to be able to modify parameters inside scoped api keys, which then defeats their point. 

Now, I can see one use-case where if a scoped API Key mentions include_fields=x,y,z and a search parameter mentions include_fields=x, we could allow it to override, since x is a subset of the fields mentioned inside the embedded parameters. But that then complicates the behavior of this feature. 

That said, you can already achieve this by generating a new scoped search API Key with include_fields=x dynamically and use that key. And since scoped api keys are ephemeral and cryptographically generated, you can create as many as you need on the fly.

**Comment by isnigurova-sc** [comment]
But we don't want to generate a scoped key for every request and it would we very useful to have this feature.

**Comment by andrey-kramarev-ocl** [comment]
Hi @jasonbosco,

Thanks for your explanation. Typesense is a great tool, and we’re looking to integrate it into our setup. From what I understand, a scoped API key is mainly for access control, where the application sets certain boundaries, and search parameters are used to narrow the results.

In our setup, we plan to generate a scoped key for each user. For example, the user could access up to 50 fields from a project based on some access rights, and then apply additional filters depending on the page they are on, retrieving only the 3 fields they need. This would be useful, particularly for multi-search, but even for simple searches it would help reduce the amount of data we transmit without needing to generate a new scoped key for every request.

Or, is there another way to limit the number of fields returned without implementing extra filtering on our end?

**Comment by isnigurova-sc** [comment]
Hi @jasonbosco,
Do you plan to implement this feature?

**Comment by jasonbosco** [comment]
> From what I understand, a scoped API key is mainly for access control, where the application sets certain boundaries, and search parameters are used to narrow the results.

This is true of the filter_by parameter. The filter_by specified inside the scoped API key is ANDed with any filters specified in the standard filter_by parameter. 

But this is not true of other search parameters. For these other params, the value inside the scoped api key overrides the value specified in the standard search parameters. 

That said, I can definitely see the utility of allowing a subset of fields from the include_fields parameter embedded inside a scoped API key to be overridden in the search parameters. 

So I'll leave this issue open as a feature request. We prioritize requests based on the factors mentioned [here](https://github.com/typesense/typesense/issues/731#issue-1381643124). 

In the meantime, one work-around could be to generate a scoped API key for each combination of include_fields on the server-side and re-using them for a user if the same set of fields are requested. This way you're not having to generate a scoped API key for each request, instead it's only once per combination of fields inside include_fields.

**Comment by Sttr69** [comment]
****
